hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -1996,7 +1996,12 @@ private void metaSave(PrintWriter out) {
     LightWeightHashSet<Long> openFileIds = new LightWeightHashSet<>();
     for (DatanodeDescriptor dataNode :
         blockManager.getDatanodeManager().getDatanodes()) {
-      for (long ucFileId : dataNode.getLeavingServiceStatus().getOpenFiles()) {
+      // Sort open files
+      LightWeightHashSet<Long> dnOpenFiles =
+          dataNode.getLeavingServiceStatus().getOpenFiles();
+      Long[] dnOpenFileIds = new Long[dnOpenFiles.size()];
+      Arrays.sort(dnOpenFiles.toArray(dnOpenFileIds));
+      for (Long ucFileId : dnOpenFileIds) {
         INode ucFile = getFSDirectory().getInode(ucFileId);
         if (ucFile == null || ucFileId <= prevId ||
             openFileIds.contains(ucFileId)) {
```

Comment:
@aajisaka Thanks for comments. Previous behavior is that inode ids are sorted for each datanode
Here the  if clause (ucFileId <= prevId) is the key point, , it affects remote iterator, if they are not sorted, some open files maybe are missing. This function getFilesBlockingDecom is added by HDFS-11847

Reply:
Agreed. It must be sorted.

I think HDFS-11847 has a bug. If the DataNodes have the following open files and we want to list all the open files:

DN1: [1001, 1002, 1003, ... , 2000]
DN2: [1, 2, 3, ... , 1000] 

At first `getFilesBlockingDecom(0, "/")` is called and it returns [1001, 1002,  ... , 2000], and next `getFilesBlockingDecom(2000, "/")` is called and the open files of DN2 is missed because the last inode Id of the previous result is 2000.

Reply:
@aajisaka Thanks, you are right, it has a bug.
How about keeping the same as before that sort the open files. File a new jira for this bug?

Reply:
Yes, I'm +1 to keep the same as before. I'll file a new jira. Thanks!

Reply:
Filed HDFS-16059